### PR TITLE
Revised .NET Framework version detection article

### LINKED ISSUE #3991: Proposal to improve "How To: Determine Which .NET Framework Versions Are Installed"
[This "How To" document](https://docs.microsoft.com/en-us/dotnet/framework/migration-guide/how-to-determine-which-versions-are-installed#net_b) is repetitive, listing the same information several times in tables, but each table is slightly different! And yet, not one of the tables provides a simple table that can be easily programmed against.

For example:

1. The table [after the first example](https://docs.microsoft.com/en-us/dotnet/framework/migration-guide/how-to-determine-which-versions-are-installed#net_b) lists Dword -> Version, but puts "installed with ..." on the Version column for .Net 4.5.x and then on the Release column for .Net 4.6.x
2. The table [after the second example](https://docs.microsoft.com/en-us/dotnet/framework/migration-guide/how-to-determine-which-versions-are-installed#net_d) has the same information, but backwards, mapping Version to the value of the Release DWORD. It's relatively consistent, keeping the information about Windows versions all on the left (Version) column, and it lists all the release numbers, even though the code in the example preceding it only uses the lowest release number for each version.
3. The table [after the PowerShell example](https://docs.microsoft.com/en-us/dotnet/framework/migration-guide/how-to-determine-which-versions-are-installed#ps_a) only shows the lowest "Release" number for each version. This ends up making the PowerShell example confusing even to PowerShell scripters (see [this question on stack overflow](https://stackoverflow.com/a/3495491/8718)), because it has this partial list -- and the code example isn't complete the way the others are.

Ideally, all three tables should be formatted the way the PowerShell table is, with just a Release number and a Version number.

I assume that the OS version _doesn't matter_ --we don't need to know it to determine the version-- and it's only listed in the examples as an explanation for why there are multiple `Release` numbers for each version, but if we need a note about which version of Windows they apply to, it should be in a separate column, and it should be consistent. Here are some examples of the current text:

- installed with Windows 8.1
- installed on Windows 8, Windows 7 SP1, or Windows Vista SP2
- installed with Windows 10
- installed on all other Windows OS versions
- installed on Windows 10 Anniversary Update
- installed on all other Windows OS versions

It's not clear whether the second "all other Windows OS versions" includes Windows 10 prior to the Anniversary Update or not. I assume that it does, because the other examples seem to mean "previous versions of Windows."

It's also not clear why these tables only talk about client OSes.

I would propose borrowing the _previous_ PowerShell script from that Stack Overflow article for the documentation on the page, using a switch with all the versions in it rather than just a single test. 

```posh
    Get-ChildItem 'HKLM:\SOFTWARE\Microsoft\NET Framework Setup\NDP' -recurse |
    Get-ItemProperty -name Version,Release -EA 0 |
    Where { $_.PSChildName -match '^(?!S)\p{L}'} |
    Select PSChildName, Version, Release, @{
        name="Product"
        expression={
            $Entry = $_
            switch -regex ($Entry.Release) {
                "^$" { <# Intentionally left blank #> }
                "378389" { [Version]"4.5" }
                "378675|378758" { [Version]"4.5.1" }
                "379893" { [Version]"4.5.2" }
                "393295|393297" { [Version]"4.6" }
                "394254|394271" { [Version]"4.6.1" }
                "394802|394806" { [Version]"4.6.2" }
                "460798|460805" { [Version]"4.7" }
                "461308|461310" { [Version]"4.7.1" }
                default { 
                Write-Warning "Release '$_' of '$($Entry.PSChildName)' version $($Entry.Version) is undocumented!" 
                Write-Host "Please update script from http://bit`.ly/netversions" -Fore Cyan
                }
            }
        }
    }
```

The script that's on the page now is different, it's using a table to look up versions just because I can easily generate it from the markdown.

### LINKED ISSUE #9174: Why make getting basic information so convoluted?
Hello Microsoft.  I don't suspect that you know or care how much time people waste collectively following these instructions just to get a simple answer about a commonly installed component do you? (rhetorical question). Does the .NET team really hate users that much that they cant be bothered to provide a simple command line utility, or a single PowerShell command, or some reference in Programs and Features or something that doesn't require the arcane procedures in articles such as this. The year is almost 2019, not 1999 - why continue to waste everyone's time like this?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 2115da38-e95e-46fc-1e2d-da88ef92c6cf
* Version Independent ID: a3790e8a-0654-851b-c991-ab7382025406
* Content: [How to: Determine which .NET Framework versions are installed](https://docs.microsoft.com/en-us/dotnet/framework/migration-guide/how-to-determine-which-versions-are-installed)
* Content Source: [docs/framework/migration-guide/how-to-determine-which-versions-are-installed.md](https://github.com/dotnet/docs/blob/master/docs/framework/migration-guide/how-to-determine-which-versions-are-installed.md)
* Product: **dotnet-framework**
* GitHub Login: @rpetrusha
* Microsoft Alias: **ronpet**

### LINKED ISSUE #9267: Complexity and Confusion
I agree with those who complain about the unnecessary complexity. The Help-&gt;About in Visual Studio 2017 tells me I have version 4.7.02556 but the Powershell method tells me 4.7.1 and clrver tells me 4.0.30319, but I think there's a difference between the framework and the runtime... correct? Is there an article that describes this?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 2115da38-e95e-46fc-1e2d-da88ef92c6cf
* Version Independent ID: a3790e8a-0654-851b-c991-ab7382025406
* Content: [How to: Determine which .NET Framework versions are installed](https://docs.microsoft.com/en-us/dotnet/framework/migration-guide/how-to-determine-which-versions-are-installed#ps_a)
* Content Source: [docs/framework/migration-guide/how-to-determine-which-versions-are-installed.md](https://github.com/dotnet/docs/blob/master/docs/framework/migration-guide/how-to-determine-which-versions-are-installed.md)
* Product: **dotnet-framework**
* GitHub Login: @rpetrusha
* Microsoft Alias: **ronpet**

### PR DESCRIPTION
**Pull request by rpetrusha**
## Revised .NET Framework version detection article

Fixes #3991 
Fixes #9174 
Fixes #9267

Related to dotnet/samples#564

Review URLs:
- [How to: Determine which .NET Framework versions are installed](https://review.docs.microsoft.com/en-us/dotnet/framework/migration-guide/how-to-determine-which-versions-are-installed?branch=pr-en-us-9954)
- [.NET Framework release keys and Windows operating system versions](https://review.docs.microsoft.com/en-us/dotnet/framework/migration-guide/release-keys-and-os-versions?branch=pr-en-us-9954)

### CODE REVIEW

**Comment by rpetrusha** [comment]
Closing and reopening to begin new build with the latest version of the examples.

**Comment by Thraka** [comment]
@rpetrusha Since you are in here revamping this, want to take #9174 into consideration? I had that on my todo for next sprint. I guess you're kind of doing it for the other bug 👍

**Comment by rpetrusha** [comment]
I've added `Fixes`, @Thraka. I think that the PR addresses the stylistic complaints in #9174, although it doesn't provide a tool to detect the version.

**Comment by Thraka** [comment]
@rpetrusha Is it possible to make known that there are community provided tools to help detection? Such as https://github.com/EliteLoser/DotNetVersionLister

File: docs/framework/migration-guide/how-to-determine-which-versions-are-installed.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
 ---
 title: "How to: Determine which .NET Framework versions are installed"
-ms.date: "04/10/2018"
+ms.date: "01/11/2019"
```
Comment (BillWagner) [inline docs/framework/migration-guide/how-to-determine-which-versions-are-installed.md]: given the delay in review time 😢 , you may want to update the date to today's

**Review by BillWagner — APPROVED** [review]
Sorry for delay in getting everything reviewed @rpetrusha 

This LGTM, and is ready to :shipit:

**Pull request merged**